File: 167.two_sum_II_input_array_is_sorted/sum.go
// 167. Two sum- Input array is sorted
// Topics: 'Array', 'Binary Search', 'Two Pointers'

// Given a 1-indexed array of integers numbers that is already sorted in non-decreasing order, find two numbers such that they add up to a specific target number. Let these two numbers be numbers[index1] and numbers[index2] where 1 <= index1 < index2 <= numbers.length.

// Return the indices of the two numbers, index1 and index2, added by one as an integer array [index1, index2] of length 2.

// The tests are generated such that there is exactly one solution. You may not use the same element twice.

// Your solution must use only constant extra space.

// Example 1:

// Input: numbers = [2,7,11,15], target = 9
// Output: [1,2]
// Explanation: The sum of 2 and 7 is 9. Therefore, index1 = 1, index2 = 2. We return [1, 2].

// Example 2:

// Input: numbers = [2,3,4], target = 6
// Output: [1,3]
// Explanation: The sum of 2 and 4 is 6. Therefore index1 = 1, index2 = 3. We return [1, 3].

// Example 3:

// Input: numbers = [-1,0], target = -1
// Output: [1,2]
// Explanation: The sum of -1 and 0 is -1. Therefore index1 = 1, index2 = 2. We return [1, 2].

// Constraints:

//     2 <= numbers.length <= 3 * 104
//     -1000 <= numbers[i] <= 1000
//     numbers is sorted in non-decreasing order.
//     -1000 <= target <= 1000
//     The tests are generated such that there is exactly one solution.

package twosumiiinputarrayissorted

func twoSum(numbers []int, target int) []int {
	for i, number := range numbers {
		t := target - number
		found := search(numbers, i, t, 0, len(numbers)-1)
		if found == -1 {
			continue
		}
		if i < found {
			return []int{i + 1, found + 1}
		}
		return []int{found + 1, i + 1}
	}
	return []int{}
}

func search(numbers []int, index, t, s, e int) int {
	for s <= e {
		m := s + (e-s)/2
		if numbers[m] == t && index != m {
			return m
		}
		if numbers[m] < t {
			s = m + 1
		} else {
			e = m - 1
		}
	}
	return -1
}

File: 1214.two_sum_BSTs/sum.py
# 1214. Two Sum BSTs
# Topics: 'Two Pointers', 'Binary Search', 'Stack', 'Tree', 'Depth-First Search', 'Binary Search Tree', 'Binary Tree'
# Level: 'Medium'

# Given the roots of two binary search trees, root1 and root2, return true if and only if there is a node in the first tree and a node in the second tree whose values sum up to a given integer target.

# Example 1:

# Input: root1 = [2,1,4], root2 = [1,0,3], target = 5

# Output: true

# Explanation: 2 and 3 sum up to 5.

# Example 2:

# Input: root1 = [0,-10,10], root2 = [5,1,7,0,2], target = 18

# Output: false


# Constraints:

#     The number of nodes in each tree is in the range [1, 5000].
#     -10⁹ <= Node.val, target <= 10⁹

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def twoSumBSTs(self, root1: Optional[TreeNode], root2: Optional[TreeNode], target: int) -> bool:
        
        x = set()
        def dfs(root):
            if not root:
                return
            x.add(target - root.val)
            dfs(root.left)
            dfs(root.right)
        
        dfs(root1)

        def search(root):
            if not root:
                return False
            if root.val in x:
                return True
            if search(root.left):
                return True
            if search(root.right):
                return True
            return False
        
        return search(root2)